src/mempool.cpp

```diff
@@ -0,0 +1,276 @@
+#include <algorithm>
+#include <system_error>
+#include <unordered_set>
+
+#include "common/error.h"
+#include "common/expect.h"
+#include "crypto/hash.h"
+#include "cryptonote_basic/cryptonote_format_utils.h"
+#include "error.h"
+#include "misc_log_ex.h"
+#include "rpc/daemon_messages.h" // external/monero/src
+#include "rpc/daemon_zmq.h"
+#include "rpc/json.h"
+#include "util/ownership_test.h"
+
+namespace lws
+{
+  std::chrono::seconds send_timeout {3};
+  std::chrono::seconds receive_timeout {5};
+
+  struct mempool::thread_detail
+  {
+    mempool& self;
+    rpc::context &context;
+    rpc::client client;
+    std::atomic<bool> stopped;
+
+    thread_detail(mempool& self, rpc::context& ctx, rpc::client&& rpc_client)
+      : self(self)
+      , context(ctx)
+      , client(std::move(rpc_client))
+      , stopped(false)
+    {}
+
+    void run()
+    {
+      while (!stopped.load(std::memory_order_acquire) && client)
+      {
+        auto pub = client.wait_for_pub();
+        if (!pub)
+        {
+          if (pub.matches(std::errc::interrupted) || pub.matches(std::errc::invalid_argument))
+            break;
+          if (pub == lws::error::daemon_timeout || pub == net::zmq::make_error_code(EAGAIN))
+          {
+            auto synced = sync_pool();
+            if (!synced)
+              MERROR("Unable to sync mempool: " << synced.error());
+            continue;
+          }
+          MERROR("Mempool loop failed: " << pub.error().message());
+          continue;
+        }
+
+        if (stopped.load(std::memory_order_acquire) || !client)
+          break;
+
+        switch(pub->first)
+        {
+          case rpc::client::topic::txpool:
+            handle_txpool(std::move(pub->second));
+            break;
+
+          case rpc::client::topic::block:
+            handle_block(std::move(pub->second));
+            break;
+        }
+      }
+    }
+
+    void handle_block(std::string&& json) {
+      const auto parsed = rpc::minimal_chain_pub::from_json(std::move(json));
+      if (!parsed)
+      {
+        MERROR("Unable to parse blockchain notification: " << parsed.error());
+        return;
+      }
+      MDEBUG("Mempool block " <<  parsed->top_block_height);
+
+      auto synced = sync_pool();
+      if (!synced)
+        MERROR("Unable to sync mempool: " << synced.error());
+    }
+
+    void handle_txpool(std::string&& json) {
+      const auto parsed = rpc::full_txpool_pub::from_json(std::move(json));
+      if (!parsed)
+      {
+        MERROR("Failed parsing txpool pub: " << parsed.error().message());
+        return;
+      }
+
+      const std::lock_guard<std::mutex> lock{self.mutex_};
+      auto next = std::make_shared<mempool::txs>(*self.current_);
+      next->reserve(parsed->txes.size());
+      for (auto& tx: parsed->txes)
+      {
+        MDEBUG("Mempool tx " << get_transaction_hash(tx) << " size: " << next->size());
+        auto shared_tx = std::make_shared<const cryptonote::transaction>(std::move(tx));
```

Comment:
Why the `shared_ptr` to the transaction, instead of just having the tx itself in the vector?